### DISCUSSION TITLE
How to best toggle an open/expanded state for nested components?

### DISCUSSION THREAD

**Discussion by HandsomeStrife** [livewire/livewire#3496] [1. Help]
Hi, I'm hoping someone can help me understand what I'm doing wrong, or some understanding of if I should be doing this differently.

Effectively, in each child component there is a modal, and on clicking a button I want the modal to open on the page - and then on clicking a button on a different child component, I wish the existing open modal to close and the new one to open. I can get this working using an `emit` function, but it has to reload and re-render every component on the page which is very slow.

As such I thought I could create a parent component that would receive the open call, and mark the child as being expanded or not. Then, on clicking a button, the child could `emitUp` the notification that it's been opened, and then the parent change which one is expanded. However, I get the error:

`Livewire encountered corrupt data when trying to hydrate the [open-connections-list] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests.`

I've also tried passing a boolean value through, but doesn't seem to work at all - doing: `:expanded="{{ $expand_id === $user_id ? 'true' : 'false' }}"` returns me an error that says `syntax error, unexpected '<'`, and doing it without the : at the start doesn't load any components at all.

I'm just not sure the best way to achieve this. It could well be that I should just be writing a bit of vanilla JS to toggle the classes on button press, but I was attempting to embrace Livewire and accomplish this without having to write extra JS.

Any help/advice would be hugely appreciated.

Here are my components:

OpenConnections.php
```
class OpenConnectionsList extends Component
{
    public $expand_id = false;
    public $connections; // Laravel Collection

    protected $listeners = ['expand'];

    public function mount($connections)
    {
        $this->connections = $connections;
    }

    public function render()
    {
        return view('livewire.open-connections-list');
    }

    public function expand($userID)
    {
        $this->expand_id = $userID;
    }
}
```

open-connections-list.blade.php
```
<div>
    @foreach ($connections as $user_id => $user_connections)
        <livewire:open-connections
            :userId="$user_id"
            :userConnections="$user_connections"
            :expandedId="$expand_id"
            :wire:key="$user_id">
    @endforeach
</div>
```

OpenConnections.php
```
class OpenConnections extends Component
{
    public $user_id;
    public $user_connections;
    public $expanded = false;

    public function mount($userId, $userConnections, $expandedId = false)
    {
        $this->user_id = $userId;
        $this->user_connections = $userConnections;
        $this->expanded = ($expandedId === $this->user_id);
    }

    public function render()
    {
        return view('livewire.open-connections', ['expanded' => $this->expanded]);
    }

    public function toggleExpanded()
    {
        $this->emitUp('expand', $this->user_id);
    }
}
```

open-connections.blade.php (partial, with a lot of the redundant HTML removed)
```
<div>
    <span>(<a href="#" wire:click="toggleExpanded">{{$expanded ? 'Hide' : 'Show'}}</a>)</span>
    <div class="sidebar-element {{ $expanded ? 'expanded' : ''  }}">
                ...
    </div>
</div>
```

**Comment by joshhanley** [comment]
@danives did you manage to sort this out?

**Reply by HandsomeStrife** [reply]
No, I ended up reverting out all the livewire aspects and doing it in vanilla JS as I kept getting hydration errors. It's a shame as I really thought it should be feasible.

**Reply by joshhanley** [reply]
@danives sorry to hear that 🙁

Well if you do decide to give it another crack, there is a couple of things I can see in what you have shared.

For future reference, this `:expanded="{{ $expand_id === $user_id ? 'true' : 'false' }}"` isn't a valid statement, so Laravel's blade compiler (which Livewire uses) will throw an error (like you got).

If you have the `:` at the front, it tells blade that you want to execute everything as php code, so it would need to be written like this `:expanded="$expand_id === $user_id ? true : false"`. But you can also shorten that down further to `:expanded="$expand_id === $user_id"` which would then pass `true` or `false` into the component (rather than the strings true or false).

For the hydration error, that can be tough to troubleshoot at times. I recently wrote a thread on twitter about what causes it, it's here if you're interested in having a look https://twitter.com/_joshhanley/status/1420919521628475393?s=20

I also recently wrote a small packed to try and assist with diagnosing this issue. It's experimental but should work, which is here https://github.com/joshhanley/livewire-checksum-logger

Hope that helps if you do run into that issue again!

**Reply by HandsomeStrife** [reply]
Hey, thanks for that - I'll give that package a try and have a look through your Twitter thread.

The `:expanded` use was mainly because I wanted to provide a boolean value through as the variable, but for whatever reason if I provided a variable that was a bool type, it wouldn't register the variable being passed through at all, with or without the curly braces (and with/without the quotes), but it might be that my mixture of uses ended up obfuscating the actual solution, so I'll give it another try. It's why I switched to passing through the ID instead, and doing the comparison in the `OpenConnections` component instead, but still ended up with the hydration problem.

I'll find the time to try and rebuild the functionality in an example repo - all being well it'll (hopefully) work, but in case it doesn't at least I'll have a code reference for anyone who wanted to dig in further :) Appreciate the response and help!

**Reply by joshhanley** [reply]
@danives no worries 🙂

Ah yeah fair enough.

If you do, let me know and I'll have a look 🙂
